core/trino-main/src/main/java/io/trino/execution/scheduler/FixedCountNodeAllocatorService.java

```diff
@@ -0,0 +1,274 @@
+            if (nodes.isEmpty()) {
+                throw new TrinoException(NO_NODES_AVAILABLE, "No nodes available to run query");
+            }
+
+            List<InternalNode> nodesMatchingRequirements = nodes.stream()
+                    .filter(node -> requirements.getAddresses().isEmpty() || requirements.getAddresses().contains(node.getHostAndPort()))
+                    .collect(toImmutableList());
+
+            if (nodesMatchingRequirements.isEmpty()) {
+                throw new TrinoException(NO_NODES_AVAILABLE, "No nodes available to run query");
+            }
+
+            Optional<InternalNode> selectedNode = nodesMatchingRequirements.stream()
+                    .filter(node -> allocationCountMap.getOrDefault(node, 0) < maximumAllocationsPerNode)
+                    .min(comparing(node -> allocationCountMap.getOrDefault(node, 0)));
+
+            if (selectedNode.isEmpty()) {
+                return Optional.empty();
+            }
+
+            allocationCountMap.compute(selectedNode.get(), (key, value) -> value == null ? 1 : value + 1);
+            return selectedNode;
+        }
+
+        private synchronized void releaseNodeInternal(InternalNode node)
+        {
+            int allocationCount = allocationCountMap.compute(node, (key, value) -> value == null ? 0 : value - 1);
+            checkState(allocationCount >= 0, "allocation count for node %s is expected to be greater than or equal to zero: %s", node, allocationCount);
+        }
+
+        private void processPendingAcquires()
+        {
+            verify(!Thread.holdsLock(this));
+
+            IdentityHashMap<PendingAcquire, InternalNode> assignedNodes = new IdentityHashMap<>();
+            IdentityHashMap<PendingAcquire, RuntimeException> failures = new IdentityHashMap<>();
+            synchronized (this) {
+                Iterator<PendingAcquire> iterator = pendingAcquires.iterator();
+                while (iterator.hasNext()) {
+                    PendingAcquire pendingAcquire = iterator.next();
+                    if (pendingAcquire.getFuture().isCancelled()) {
+                        iterator.remove();
+                        continue;
+                    }
+                    try {
+                        Optional<InternalNode> node = tryAcquireNode(pendingAcquire.getNodeRequirements());
+                        if (node.isPresent()) {
+                            iterator.remove();
+                            assignedNodes.put(pendingAcquire, node.get());
+                        }
+                    }
+                    catch (RuntimeException e) {
+                        iterator.remove();
+                        failures.put(pendingAcquire, e);
+                    }
+                }
+            }
+
+            assignedNodes.forEach((pendingAcquire, node) -> {
+                SettableFuture<InternalNode> future = pendingAcquire.getFuture();
+                future.set(node);
+                if (future.isCancelled()) {
```

Comment:
This is violation of contract. The part which calls `acquire` is responsible for calling `release` on obtained lease.
We are looking at `cancelled` state only because future is being cancelled as part of `FixedCountNodeLease.release()`.
We are not expecting external user to cancel future directly.

Reply:
> We are looking at cancelled state only because future is being cancelled as part of FixedCountNodeLease.release().
We are not expecting external user to cancel future directly.

This is part of some next commit?, Because I don't see it in this commit.

Regarding (`FixedCountNodeLease.release()`):
```
                if (released.compareAndSet(false, true)) {
                    node.cancel(true);
                    if (node.isDone() && !node.isCancelled()) {
                        releaseNode(getFutureValue(node));
                    }
                }
```
that also doesn't seem right:
1. you call `node.cancel(true);` and then immediately later check `!node.isCancelled()`. I think it's always false (for `SettableFuture` at least). Do you lack test case?
2. If you remove `!node.isCancelled()` then you are left with:
```
                    node.cancel(true);
                    if (node.isDone()) {
                        releaseNode(getFutureValue(node));
                    }
```
which is prone to a race I've mentioned here (`releaseNode(getFutureValue(node))` might never be called)

I think to make it work correctly you should actually check `boolean` flags returned by `Future#set` and `Future#cancel`, so:
```
                SettableFuture<InternalNode> future = pendingAcquire.getFuture();
                if (!future.set(node)) {
                  releaseNodeInternal(node);
                }
```

and in `FixedCountNodeLease.release()` you probably should do:
```
                if (released.compareAndSet(false, true)) {
                    if (!node.cancel(true)) && node.isDone()) {
                        releaseNode(getFutureValue(node));
                    }
                }
```
I'm not 100% certain though if you `future#set` can be async (e.g. cause both `Future#cancel` and `Future#isDone` to return `false`).